src/Database/Statement.php

```diff
@@ -0,0 +1,399 @@
+    protected function performBind(string|int $column, mixed $value, int $type): void
+    {
+        $this->statement->bindValue($column, $value, $type);
+    }
+
+    /**
+     * Executes the statement by sending the SQL query to the database. It can optionally
+     * take an array or arguments to be bound to the query variables. Please note
+     * that binding parameters from this method will not perform any custom type conversion
+     * as it would normally happen when calling `bindValue`.
+     *
+     * @param array|null $params list of values to be bound to query
+     * @return bool true on success, false otherwise
+     */
+    public function execute(?array $params = null): bool
+    {
+        $this->rows = [];
+        $this->rowsFetched = false;
+        $this->executed = true;
+
+        $result = $this->statement->execute($params);
+        if ($result && $this->logger) {
+            //$this->logger->debug();
+        }
+
+        return $result;
+    }
+
+    /**
+     * Closes a cursor in the database, freeing up any resources and memory
+     * allocated to it. In most cases you don't need to call this method, as it is
+     * automatically called after fetching all results from the result set.
+     *
+     * @return void
+     */
+    public function closeCursor(): void
+    {
+        $this->statement->closeCursor();
+    }
+
+    /**
+     * Returns the next row for the result set after executing this statement.
+     * Rows can be fetched to contain columns as names or positions. If no
+     * rows are left in result set, this method will return false.
+     *
+     * ### Example:
+     *
+     * ```
+     * $statement = $connection->prepare('SELECT id, title from articles');
+     * $statement->execute();
+     * print_r($statement->fetch('assoc')); // will show ['id' => 1, 'title' => 'a title']
+     * ```
+     *
+     * @param int $mode Fetch mode. PDO::FETCH_* constant or name.
+     * @return mixed
+     */
+    public function fetch(string|int $mode = 'num'): mixed
+    {
+        $mode = static::MODES[$mode] ?? $mode;
+        if (is_string($mode)) {
+            throw new InvalidArgumentException(sprintf(
+                'Invalid named fetch mode `%s` requested. Must be one of %s.',
+                $mode,
+                array_keys(static::MODES)
+            ));
+        }
+
+        if ($this->bufferRows && $this->rowsFetched) {
+            // Can we validate the fetch mode matches? It might be row-specific.
+            $row = next($this->rows);
+            if ($row === false) {
+                return false;
+            }
+
+            foreach ($this->modifiers as $modifier) {
+                $row = $modifier($row);
+            }
+
+            return $row;
+        }
+
+        $row = $this->statement->fetch($mode);
+        if ($row === false) {
+            $this->rowsFetched = true;
+
+            return false;
+        }
+
+        if ($this->bufferRows) {
+            $this->rows[] = $row;
```

Comment:
Why not cache the rows after running the modifiers?

Reply:
The association code is madness and will fully fetch and iterate through all rows before resetting and attaching a modifier.